.travis.yml

```diff
@@ -9,24 +8,54 @@ language: go
 cache:
   directories:
     - .cache
+    - $HOME/.build
     # cache the Go module cache
     - $HOME/gopath/pkg/mod
+    # cache `go install`-ed binaries across stages. build.sh unsets GOPATH
+    # so this is hitting the default go env of $HOME/go
+    - $HOME/go/bin
+
+stages:
+  - setup
+  - lint
+  - test
+  - build release
+  - publish images
 
 jobs:
   include:
+    - stage: setup
+      script:
+        - ./build.sh setup
+        - ./build.sh utils
+      os: macos
+    - stage: setup
+      script:
+        - ./build.sh setup
+        - ./build.sh utils
+      os: linux
+      dist: xenial
     - stage: lint
       script:
         - shellcheck build.sh
+    - stage: test
+      script:
+        - ./build.sh race-test
+      os: macos
+    - stage: test
+      script:
+        - ./build.sh race-test
+      os: linux
+      dist: xenial
     - stage: build release
       script:
-        - ./build.sh setup
-        - ./build.sh utils
         - ./build.sh protobuf_verify
```

Comment:
I wonder if we should move this into the lint stage?

Reply:
great idea.